src/openfermion/utils/_unitary_cc.py

```diff
@@ -0,0 +1,199 @@
+    return single_amplitudes_list, double_amplitudes_list
+
+
+def uccsd_singlet_paramsize(n_qubits, n_electrons):
+    """Determine number of independent amplitudes for singlet UCCSD
+
+    Args:
+        n_qubits(int): Number of qubits/spin-orbitals in the system
+        n_electrons(int): Number of electrons in the reference state
+
+    Returns:
+        Number of independent parameters for singlet UCCSD with a single
+        reference.
+    """
+    n_occupied = int(numpy.ceil(n_electrons / 2.))
+    n_virtual = n_qubits / 2 - n_occupied
+
+    n_single_amplitudes = n_occupied * n_virtual
+    n_double_amplitudes = n_single_amplitudes ** 2
+    return (n_single_amplitudes + n_double_amplitudes)
+
+
+def uccsd_singlet_operator(packed_amplitudes,
+                           n_qubits,
+                           n_electrons):
+    """Create a singlet UCCSD generator for a system with n_electrons
+
+    This function generates a FermionOperator for a UCCSD generator designed
+        to act on a single reference state consisting of n_qubits spin orbitals
+        and n_electrons electrons, that is a spin singlet operator, meaning it
+        conserves spin.
+
+    Args:
+        packed_amplitudes(ndarray): Compact array storing the unique single
+            and double excitation amplitudes for a singlet UCCSD operator.
+            The ordering lists unique single excitations before double
+            excitations.
+        n_qubits(int): Number of spin-orbitals used to represent the system,
+            which also corresponds to number of qubits in a non-compact map.
+        n_electrons(int): Number of electrons in the physical system.
+
+    Returns:
+        uccsd_generator(FermionOperator): Generator of the UCCSD operator that
+            builds the UCCSD wavefunction.
+    """
+    n_occupied = int(numpy.ceil(n_electrons / 2.))
+    n_virtual = int(n_qubits / 2 - n_occupied)  # Virtual Spatial Orbitals
+    n_t1 = int(n_occupied * n_virtual)
+
+    t1 = packed_amplitudes[:n_t1]
+    t2 = packed_amplitudes[n_t1:]
+
+    def t1_ind(i, j):
+        return i * n_occupied + j
+
+    def t2_ind(i, j, k, l):
+        return (i * n_occupied * n_virtual * n_occupied +
+                j * n_virtual * n_occupied +
+                k * n_occupied +
+                l)
+
+    uccsd_generator = FermionOperator()
+
+    spaces = range(n_virtual), range(n_occupied), range(2)
+
+    for i, j, s in itertools.product(*spaces):
+        uccsd_generator += FermionOperator(
+            (
+                (2 * (i + n_occupied) + s, 1),
+                (2 * j + s, 0),
+            ),
+            coefficient=t1[t1_ind(i, j)])
+
+        uccsd_generator += FermionOperator(
+            (
+                (2 * j + s, 1),
+                (2 * (i + n_occupied) + s, 0),
+            ),
+            coefficient=-t1[t1_ind(i, j)])
+
+    for i, j, s, i2, j2, s2 in itertools.product(*spaces, repeat=2):
```

Comment:
Can you add a couple single-line comments to the function to explain what you are doing in blocks of code such as this one?

Reply:
Added some description to the iterator space